### NCBI

[NCBI prokaryotic genome annotation pipeline.](https://www.ncbi.nlm.nih.gov/pubmed/27342282)  
Tatusova T, DiCuccio M, Badretdin A, Chetvernin V, Nawrocki EP, Zaslavsky L, Lomsadze A, Pruitt KD, Borodovsky M, Ostell J.  
Nucleic Acids Res. 2016 Aug 19;44(14):6614-24. Epub 2016 Jun 24.

[RefSeq: an update on prokaryotic genome annotation and curation.](https://www.ncbi.nlm.nih.gov/pubmed/29112715)  
Haft DH, DiCuccio M, Badretdin A, Brover V, Chetvernin V, O'Neill K, Li W, Chitsaz F, Derbyshire MK, Gonzales NR, Gwadz M, Lu F, Marchler GH, Song JS, Thanki N, Yamashita RA, Zheng C, Thibaud-Nissen F, Geer LY, Marchler-Bauer A, Pruitt KD.  
Nucleic Acids Res. 2018 Jan 4;46(D1):D851-D860.

[Using average nucleotide identity to improve taxonomic assignments in prokaryotic genomes at the NCBI.](https://www.ncbi.nlm.nih.gov/pubmed/29792589)  
Ciufo S, Kannan S, Sharma S, Badretdin A, Clark K, Turner S, Brover S, Schoch CL, Kimchi A, DiCuccio M.  
Int J Syst Evol Microbiol. 2018 Jul;68(7):2386-2392.

[SKESA: strategic k-mer extension for scrupulous assemblies.](https://www.ncbi.nlm.nih.gov/pubmed/30286803)  
Alexandre Souvorov, Richa Agarwala and David J. Lipman  
Genome Biol. 2018 Oct 4;19(1):153.
### GeneMarkS-2+

[Modeling leaderless transcription and atypical genes results in more accurate gene prediction in prokaryotes.](https://www.ncbi.nlm.nih.gov/pubmed/29773659/)  
Lomsadze A, Gemayel K, Tang S, Borodovsky M.  
Genome Research. 2018; 28(7):1079-1089.
### TIGRFAMs

[TIGRFAMs: a protein family resource for the functional identification of proteins.](https://www.ncbi.nlm.nih.gov/pubmed/11125044)  
Haft DH, Loftus BJ, Richardson DL, Yang F, Eisen JA, Paulsen IT, White O.  
Nucleic Acids Res. 2001 Jan 1;29(1):41-3.

[The TIGRFAMs database of protein families.](https://www.ncbi.nlm.nih.gov/pubmed/12520025)  
Haft DH, Selengut JD, White O.  
Nucleic Acids Res. 2003 Jan 1;31(1):371-3.

[TIGRFAMs and Genome Properties: tools for the assignment of molecular function and biological process in prokaryotic genomes.](https://www.ncbi.nlm.nih.gov/pubmed/17151080)  
Selengut JD, Haft DH, Davidsen T, Ganapathy A, Gwinn-Giglio M, Nelson WC, Richter AR, White O.  
Nucleic Acids Res. 2007 Jan;35(Database issue):D260-4. Epub 2006 Dec 6.

[TIGRFAMs and Genome Properties in 2013.](https://www.ncbi.nlm.nih.gov/pubmed/23197656)  
Haft DH, Selengut JD, Richter RA, Harkins D, Basu MK, Beck E.  
Nucleic Acids Res. 2013 Jan;41(Database issue):D387-95. doi: 10.1093/nar/gks1234. Epub 2012 Nov 28.
### RFAMs

[Rfam 12.0: updates to the RNA families database.](https://www.ncbi.nlm.nih.gov/pubmed/25392425)  
Nawrocki EP, Burge SW, Bateman A, Daub J, Eberhardt RY, Eddy SR, Floden EW, Gardner PP, Jones TA, Tate J, Finn RD.  
Nucleic Acids Res. 2015 Jan;43(Database issue):D130-7. doi: 10.1093/nar/gku1063. Epub 2014 Nov 11.
### PILER-CR

[PILER-CR: fast and accurate identification of CRISPR repeats.](https://www.ncbi.nlm.nih.gov/pubmed/17239253)  
Edgar RC.  
BMC Bioinformatics. 2007 Jan 20;8:18.
### CRISPR Recognition Tool (CRT)

[CRISPR recognition tool (CRT): a tool for automatic detection of clustered regularly interspaced palindromic repeats.](https://www.ncbi.nlm.nih.gov/pubmed/17577412)  
Bland C, Ramsey TL, Sabree F, Lowe M, Brown K, Kyrpides NC, Hugenholtz P.  
BMC Bioinformatics. 2007 Jun 18;8:209.
### tRNAscan-SE

[tRNAscan-SE: Searching for tRNA Genes in Genomic Sequences.](https://www.ncbi.nlm.nih.gov/pubmed/31020551)  
Chan PP, Lowe TM.  
Methods Mol Biol. 2019;1962:1-14. doi: 10.1007/978-1-4939-9173-0_1

